DataFormats/TrackerRecHit2D/interface/VectorHit.h

```diff
@@ -0,0 +1,168 @@
+#ifndef DataFormats_TrackerRecHit2D_VectorHit_h
+#define DataFormats_TrackerRecHit2D_VectorHit_h
+
+/** \class VectorHit
+ *
+ * 4-parameter RecHits for Phase2 Tracker (x,y, dx/dz, dy/dz)
+ *
+ * $Date: 2015/03/30 $
+ * \author Erica Brondolin
+ *
+ */
+
+#include "DataFormats/TrackerRecHit2D/interface/BaseTrackerRecHit.h"
+#include "DataFormats/TrackerRecHit2D/interface/VectorHit2D.h"
+#include "DataFormats/TrackerRecHit2D/interface/OmniClusterRef.h"
+
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
+
+#include "DataFormats/Phase2TrackerCluster/interface/Phase2TrackerCluster1D.h"
+#include "DataFormats/GeometryVector/interface/LocalVector.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+
+#include "DataFormats/TrackingRecHit/interface/KfComponentsHolder.h"
+
+#include "DataFormats/TrackerRecHit2D/interface/TkCloner.h"
+
+class VectorHit final : public BaseTrackerRecHit {
+public:
+  typedef OmniClusterRef::Phase2Cluster1DRef ClusterRef;
+
+  VectorHit() : thePosition(), theDirection(), theCovMatrix(), theDimension(0) { setType(bad); }
+
+  VectorHit(const VectorHit& vh);
+
+  VectorHit(const GeomDet& idet,
+            const LocalPoint& posInner,
+            const LocalVector& dir,
+            const std::array<std::array<float, 4>, 4> covMatrix,
```

Comment:
This used to be a `AlgebraicSymMatrix44` until [https://github.com/cms-sw/cmssw/commit/b6b9e9e1341e9694e8d6a5fdab07430eead966b2](b6b9e9e). But constructing new `AlgebraicSymMatrix44` in the copy constructor of `VectorHits` turned out to be very slow and accounted for the majority of the added CPU time of the new code. To avoid it, I switched to the array of array for the internal storage of covariance matrix.

Reply:
the commit points to 
```diff
-  AlgebraicSymMatrix theCovMatrix;
+  std::array<std::array<float, 4>, 4> theCovMatrix;
```
that was not AlgebraicSymMatrix44 (root SMatrix), that was the known super-slow CLHEP AlgebraicSymMatrix

Reply:
Ah, true, I misremembered that, and wasn't aware of the difference in the first place. I can check the timing with AlgebraicSymMatrix44 to see if we can use it instaed of the arrays.

Reply:
AlgebraicSymMatrix44 is 15 doubles

Reply:
right, but it's exactly the type used in the KFComponents, which should make it more effective for copying/using there.
I can see how double vs float can matter at the level of persistence.
Are we going to save VectorHits collections on file?

Reply:
Ah, another complication is that `paramsError` has to return a `AlgebraicSymMatrix` as this is the type it returns in the base class. So while we could internally store as `AlgebraicSymMatrix44`, we would always have to convert in the getter function.

Reply:
did you mean `  AlgebraicSymMatrix parametersError() const override;`? I propose to remove overload of this method in this class, it seems not necessary and is not used in other phase2 rechit classes.

Reply:
I switch this all to `AlgebraicSymMatrix44` now, that is certainly much faster than `AlgebraicSymMatrix`. I removed the overload according to this suggestion.